host/programexecutor.go

```diff
@@ -0,0 +1,574 @@
+func (pe *programExecutor) FinalizeContract(req rhp.RPCFinalizeProgramRequest) (types.FileContractRevision, error) {
+	current := pe.contract.Revision
+	rev := pe.contract.Revision
+	rev.RevisionNumber = req.NewRevisionNumber
+	req.NewOutputs.Apply(&rev)
+	// update the contract's merkle root and file size.
+	rev.FileMerkleRoot = pe.newMerkleRoot
+	rev.Filesize = pe.newFileSize
+
+	// validate that the renter's revision is valid and only transfers the
+	// additional collateral and storage costs to the void. All other
+	// costs have already been paid by the RPC budget.
+	if err := validateProgramRevision(current, rev, pe.additionalStorage, pe.additionalCollateral); err != nil {
+		return types.FileContractRevision{}, fmt.Errorf("failed to verify contract revision: %w", err)
+	}
+
+	// validate the renter's signature
+	signHash := pe.vc.ContractSigHash(rev)
+	if !pe.contract.Parent.RenterPublicKey.VerifyHash(signHash, req.Signature) {
+		return types.FileContractRevision{}, errors.New("renter signature invalid")
+	}
+
+	pe.contract.Revision = rev
+	pe.contract.HostSignature = types.SignHash(pe.host.privkey, signHash)
+	pe.contract.RenterSignature = req.Signature
+
+	return pe.contract.FileContractRevision, nil
+}
+
+// Revert removes the sectors that were added by the program. If
+// commit has already been called, this function is a no-op.
+func (pe *programExecutor) Revert() error {
+	if pe.committed {
+		return nil
+	}
+
+	// delete the gained sectors.
+	for root := range pe.gainedSectors {
+		if err := pe.host.sectors.DeleteSector(root); err != nil {
+			return fmt.Errorf("failed to remove sector: %w", err)
+		}
+	}
+
+	// increase the budget by the failure refund. This will refund the storage
+	// costs from executing the program to the renter.
+	pe.budget.Increase(pe.failureRefund)
+	return nil
+}
+
+// Commit removes any sectors that were removed by the program and
+// sets the failure refund to zero. If commit has already been called this
+// function is a no-op.
+func (pe *programExecutor) Commit() error {
+	if pe.committed {
+		return nil
+	}
+	// reset the gained sectors.
+	pe.gainedSectors = make(map[types.Hash256]struct{})
+	// delete the removed sectors.
+	for root := range pe.removedSectors {
+		if err := pe.host.sectors.DeleteSector(root); err != nil {
+			return fmt.Errorf("failed to remove sector: %w", err)
+		}
+	}
+
+	// all program ops are now committed, set the failure refund to zero.
+	pe.failureRefund = types.ZeroCurrency
+	return nil
+}
+
+// newExecutor initializes the program's executor.
+func newExecutor(sh *SessionHandler, vc consensus.ValidationContext, settings rhp.HostSettings, budget *rpcBudget) *programExecutor {
+	pe := &programExecutor{
+		hostKey: sh.pubkey,
+
+		settings: settings,
+		budget:   budget,
+		duration: 1,
+		host:     sh,
+		vc:       vc,
+
+		gainedSectors:  make(map[types.Hash256]struct{}),
+		removedSectors: make(map[types.Hash256]struct{}),
+	}
+	pe.encoder = types.NewEncoder(&pe.output)
```

Comment:
It's pretty awkward, but I don't think that makes sense since the executor expects exclusive control of its output buffer, and `ExecuteInstruction` requires an `io.Writer` for flushing the response + buffer.